src/fs/metadata_log.cc

```diff
@@ -0,0 +1,326 @@
+            auto& curr_dir = dir_it->second.get_directory();
+
+            auto it = curr_dir.entries.find(component);
+            if (it == curr_dir.entries.end()) {
+                return path_lookup_error::NO_ENTRY;
+            }
+
+            inode_t entry_inode = it->second;
+            if (check_if_dir) {
+                auto entry_it = _inodes.find(entry_inode);
+                assert(entry_it != _inodes.end() and "dir entries have to exist");
+                if (not entry_it->second.is_directory()) {
+                    return path_lookup_error::NOT_DIR;
+                }
+            }
+
+            components_stack.emplace_back(entry_inode);
+        }
+    }
+
+    return components_stack.back();
+}
+
+future<inode_t> metadata_log::futurized_path_lookup(const sstring& path) const {
+    auto lookup_res = path_lookup(path);
+    return std::visit(overloaded {
+        [](path_lookup_error error) {
+            switch (error) {
+            case path_lookup_error::NOT_ABSOLUTE:
+                return make_exception_future<inode_t>(path_is_not_absolute());
+            case path_lookup_error::NO_ENTRY:
+                return make_exception_future<inode_t>(no_such_file_or_directory());
+            case path_lookup_error::NOT_DIR:
+                return make_exception_future<inode_t>(path_component_not_directory());
+            }
+            __builtin_unreachable();
+        },
+        [](inode_t inode) {
+            return make_ready_future<inode_t>(inode);
+        }
+    }, lookup_res);
+}
+
+file_offset_t metadata_log::file_size(inode_t inode) const {
+    auto it = _inodes.find(inode);
+    if (it == _inodes.end()) {
+        throw invalid_inode_exception();
+    }
+
+    return std::visit(overloaded {
+        [](const inode_info::file& file) {
+            return file.size();
+        },
+        [](const inode_info::directory&) -> file_offset_t {
+            throw invalid_inode_exception();
+        }
+    }, it->second.contents);
+}
+
+future<inode_t> metadata_log::create_file(sstring path, file_permissions perms) {
+    return create_file_operation::perform(*this, std::move(path), std::move(perms), false);
+}
+
+future<> metadata_log::create_directory(sstring path, file_permissions perms) {
+    return create_file_operation::perform(*this, std::move(path), std::move(perms), true).discard_result();
+}
+
+// TODO: think about how to make filesystem recoverable from ENOSPACE situation: flush() (or something else) throws ENOSPACE, then it should be possible to compact some data (e.g. by truncating a file) via top-level interface and retrying the flush() without a ENOSPACE error. In particular if we delete all files after ENOSPACE it should be successful. It becomes especially hard if we write metadata to the last cluster and there is no enough room to write these delete operations. We have to guarantee that the filesystem is in a recoverable state then.
```

Comment:
Warm congratulations on a 557 character line! 🎉 Please break it a little, since I got bored scrolling and I still don't know the last 25% of this comment :)